File: leetcodes/435. Non-overlapping Intervals/nonOverlappingIntervals.js
/**
 * @param {number[][]} intervals
 * @return {number}
 */
 var eraseOverlapIntervals = function(intervals) {
  //input: 2d arrays
  //output: integer
  //edge case: intervals.length >=1, intervals[i].length == 2
  /**
   * Greedy approach:  time O(n) = nlogn (sorting) , space O(n) = O(1)
   * Sort the array with the end time
   * set prev is interval[0]
   * iterate over the interval from i = 1 to intervals.length-1
   *   if prev[1] > intervals[i][0], cnt ++
   *   else prev = interval[i]
   *
   * return cnt
   * ex: [1,2] and [2,3] they are not overlapping
   */
  if (intervals.length === 1) return 0;
  intervals.sort((a, b) => a[1] - b[1]);
  let prev = intervals[0];
  let count = 0;
  for (let i = 1; i < intervals.length; i++) {
    if (prev[1] > intervals[i][0]) count++;
    else prev = intervals[i];
  }
  return count;
};

File: leetcodes/435. Non-overlapping Intervals/nonOverlappingIntervals.java
class Solution {
  public int eraseOverlapIntervals(int[][] intervals) {
    //The problem asks for minimum removal.
    //We only need to remove [1,100] and [2,12] to get rid of over-lapping interval
    //sort the array
    //iterate over the list, to count overlapping part
    //  else set start, end to cur index
    //time O(n) = nlogn (sorting) space O(1)
    if (intervals.length == 1) return 0;
    Arrays.sort(intervals, (o1, o2)->o1[0]-o2[0]);
    int count = 0;
    int end = intervals[0][1];
    for (int i = 1; i<intervals.length; i++) {
      if (intervals[i][0] < end) {
        count++;
        end = Math.min(intervals[i][1] , end);
      } else {
        end = intervals[i][1];
      }
    }
    return count;
  }
}